### ISSUE TITLE
Possible bug in Expansion Algorithm step 13.4.16

### ISSUE THREAD

**Issue by kasei** [w3c/json-ld-api#270]
[Expansion Algorithm](https://w3c.github.io/json-ld-api/#expansion-algorithm) step 13.4.16 says:

> Unless expanded value is null, and expanded property is `@value`, and input type is `@json`, set the expanded property entry of result to expanded value.

I believe test [t0122](https://w3c.github.io/json-ld-api/tests/expand-manifest.html#t0122) highlights a case where this is problematic. It's possible by way of 13.4.3.2 and (IRI Expansion step 2) that `expanded value` is null here, but the test expects the resulting value to be an empty string. I'm not sure where else the `null` value might be converted to an empty string. Is it possible the test is wrong? Or does the wording in 13.4.16 need updating?

**Comment by gkellogg** [comment]
You might be referring to [compact#tjs11](https://w3c.github.io/json-ld-api/tests/compact-manifest.html#tjs11), which doesn't have a related expansion test. [t0122](https://w3c.github.io/json-ld-api/tests/expand-manifest.html#t0122) is about catching IRIs having the form of a keyword. I'll add a couple of tests for this.

And, that test should read, instead:

> Unless _expanded value_ is `null`, and _expanded property_ is `@value`, and _input type_ is **not**  `@json`, set the _expanded property_ entry of _result_ to _expanded value_.

But, this step overlaps a change in PR #275.

**Comment by kasei** [comment]
I'm not doing anything related to compaction, so I doubt there's any confusion with tjs11.

**Comment by gkellogg** [comment]
Can you describe what it is about t0122 that relates to this processing step?

**Comment by gkellogg** [comment]
To be clear, [t0122](https://w3c.github.io/json-ld-api/tests/expand-manifest.html#t0122) contains the following input file:

```json
{
  "@context": {
    "@base": "http://example.org/"
  },
  "http://example.org/vocab/at": {"@id": "@"},
  "http://example.org/vocab/foo.bar": {"@id": "@foo.bar"},
  "http://example.org/vocab/ignoreme": {"@id": "@ignoreMe"}
}
```

which doesn't seem to relate to JSON literals. One test I was going to add uses the following input file:

```json
{
  "@context": {
    "@version": 1.1,
    "e": {"@id": "http://example.org/vocab#null", "@type": "@json"}
  },
  "e": null
}
```

I think this would trigger the issue you were concerned about.

**Comment by kasei** [comment]
@gkellogg 

> which doesn't seem to relate to JSON literals

Is the "unless" that begins 13.4.16 meant to apply *only* to the first clause about `expanded value` and not to the entire 3-clause conjunction?

**Comment by kasei** [comment]
My concern here is is about `@ignoreMe` turning into the empty string in [0122-out.jsonld](https://w3c.github.io/json-ld-api/tests/expand/0122-out.jsonld). My reading of 13.4.16 is that it isn't about JSON literals, as I parsed it as something like:

> "if not((`expanded value` is `null`) && (`expanded property` is `@value`) && (`input type` is `@json`))

Given that reading/parsing, I can't/couldn't figure out how to end up with an empty string as: 13.4.3.2 will set `expanded value` to `null`; then 13.4.16 is triggered (as a result of `expanded property` *not* being `@value`; it is `@id`) and `result[@id]` is set to `null`.

**Comment by gkellogg** [comment]
Note that 13.4.16 is updated by PR #275 to be "Unless _expanded value_ is `null`, and _expanded property_ is `@value`, and _input type_ is **not** `@json`, set the _expanded property_ entry of _result_ to _expanded value_."

This would be true for any non-null values of expanded value, any null values where property is not `@value`, or any null values where property is `@value` and input type is `@json`. So, it's about JSON literals when `@value` is `null` and `@type` is `@json`.

**Comment by gkellogg** [comment]
Relating to your original question (which I now finally understand), yes it does return null. It seems a bug in my code, from which the result is taken, turned it into an empty string. The test result should be updated to be the following:

```json
[{
  "http://example.org/vocab/at": [{"@id": "http://example.org/@"}],
  "http://example.org/vocab/foo.bar": [{"@id": "http://example.org/@foo.bar"}],
  "http://example.org/vocab/ignoreme": [{"@id": null}]
}]
```

**Comment by gkellogg** [comment]
@kasei I believe the corrected t0122 is correct, and I added two other tests for null JSON literals.

Because we're changing expected results, the WG will need to consider this.

**Comment by kasei** [comment]
@gkellogg updated tests looks good, thanks.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2020/2020-01-10-json-ld.html#section3-2).

- `RESOLVED:  Accept #270 including both text change to the algorithm and fix to the expected test result, noting that this would potentially affect other implementations’ test results, but that we do not believe this requires re-entry to CR`
<details><summary><i class="differentiate">View the transcript</i></summary>
Possible bug in Expansion Algorithm step 13.4.16<br/>
Rob Sanderson: link: <a href="https://github.com/w3c/json-ld-api/issues/270">https://github.com/w3c/json-ld-api/issues/270</a><br/>
Rob Sanderson: so the next one is issue 270<br/>
… after some discussion gkellogg believes the test is fixed<br/>
… and added two more tests<br/>
… but since we’re changing the tests, we do need to consider this as a group<br/>
… any change to the algorithm that doesn’t change the test results is fine<br/>
… but this one changes the test results…so we need to determine the weight of this one<br/>
… it doesn’t seem like a dropping us out of CR sort of change<br/>
… gkellogg can you elaborate?<br/>
Gregg Kellogg: there was a bug in my implementation based on my misinterpretation of my own text<br/>
… I’ll have to dig into it again more closely<br/>
… I’d misinterpreted what his concern was initially<br/>
Rob Sanderson: what I recall is that it was a null vs. empty list question<br/>
… and that the test results needed to test for whichever one it should have been rather than the other one<br/>
Gregg Kellogg: so the text had previously said if an input type was <code>@json</code><br/>
… because null would be kept only in the case for <code>@value</code> when the type is <code>@json</code><br/>
… that was a change to the algorithm text<br/>
… in terms of what the test result was…<br/>
… one of these was redundant<br/>
… expand 122 out. it had been an empty string…and should have stayed null<br/>
… so it partly related to that wrong step that shouldn’t have been in there<br/>
Rob Sanderson: so the good thing is that Greg agrees with these fixes<br/>
… so I think that we should have a resolution<br/>
… that there was essentially a typo in the test result<br/>
… and that we’re fixing it by tweaking both the algorithm and the test result<br/>
… and the issue submitter (Greg) agrees with this fix<br/>
… and that we agree this shouldn’t trigger a new CR<br/>
… my proposal, then, would be that we suggest a non-normative tweak to the algorithm text and fix what amounts to a typo in the tests<br/>
Ivan Herman: so, since as you say the algorithm is non-normative, then we should be fine<br/>
… certainly the algorithm and tests should be in sync<br/>
Rob Sanderson: pchampin you’re ok with this also?<br/>
Pierre-Antoine Champin: yes. it’s fine.<br/>
Rob Sanderson: so we should note that this would effect other implementation’s test results potentially<br/>
Proposed resolution: Accept #270 including both text change to the algorithm and fix to the expected test result, noting that this would potentially affect other implementations’ test results, but that we do not believe this requires re-entry to CR (Rob Sanderson)<br/>
Rob Sanderson: but that we do not believe this requires re-entry into CR<br/>
Gregg Kellogg: +1<br/>
Benjamin Young: +1<br/>
Jeff Mixter: +1<br/>
Tim Cole: +1<br/>
Pierre-Antoine Champin: +1<br/>
Rob Sanderson: +1<br/>
Resolution #3: Accept #270 including both text change to the algorithm and fix to the expected test result, noting that this would potentially affect other implementations’ test results, but that we do not believe this requires re-entry to CR<br/>
Ivan Herman: +1<br/>
</details>
